object/crypto_manager_kmip.go

```diff
@@ -0,0 +1,309 @@
+
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+type CryptoManagerKmip struct {
+	Common
+}
+
+func CryptoManagerKmipMoRef() types.ManagedObjectReference {
+	return types.ManagedObjectReference{
+		Type:  "CryptoManagerKmip",
+		Value: "CryptoManager",
+	}
+}
+
+func NewCryptoManagerKmip(
+	c *vim25.Client,
+	ref types.ManagedObjectReference) *CryptoManagerKmip {
+
+	return &CryptoManagerKmip{
+		Common: NewCommon(c, ref),
+	}
+}
+
+func (m CryptoManagerKmip) ListKmipServers(
+	ctx context.Context,
+	limit *int32) ([]types.KmipClusterInfo, error) {
+
+	req := types.ListKmipServers{
+		This:  m.Reference(),
+		Limit: limit,
+	}
+	res, err := methods.ListKmipServers(ctx, m.Client(), &req)
+	if err != nil {
+		return nil, err
+	}
+	return res.Returnval, nil
+}
+
+func (m CryptoManagerKmip) IsDefaultProviderNative(
+	ctx context.Context,
+	entity *types.ManagedObjectReference,
+	defaultsToParent bool) (bool, error) {
+
+	defaultProviderID, err := m.GetDefaultKmsClusterID(
+		ctx, entity, defaultsToParent)
+	if err != nil {
+		return false, err
+	}
+	if defaultProviderID == "" {
+		return false, nil
+	}
+	return m.IsNativeProvider(ctx, defaultProviderID)
+}
+
+func (m CryptoManagerKmip) IsNativeProvider(
+	ctx context.Context,
+	providerID string) (bool, error) {
+
+	info, err := m.GetClusterStatus(ctx, providerID)
+	if err != nil {
+		return false, err
+	}
+	if info == nil {
+		return false, nil
+	}
+	return info.ManagementType == string(
+		types.KmipClusterInfoKmsManagementTypeNativeProvider), nil
+}
+
+func (m CryptoManagerKmip) GetDefaultKmsClusterID(
+	ctx context.Context,
+	entity *types.ManagedObjectReference,
+	defaultsToParent bool) (string, error) {
+
+	req := types.GetDefaultKmsCluster{
+		This:             m.Reference(),
+		Entity:           entity,
+		DefaultsToParent: &defaultsToParent,
+	}
+	res, err := methods.GetDefaultKmsCluster(ctx, m.Client(), &req)
+	if err != nil {
+		return "", err
+	}
+	if res.Returnval != nil {
+		return res.Returnval.Id, nil
+	}
+	return "", nil
+}
+
+func (m CryptoManagerKmip) GetClusterStatus(
+	ctx context.Context,
+	providerID string) (*types.CryptoManagerKmipClusterStatus, error) {
+
+	req := types.RetrieveKmipServersStatus_Task{
+		This: m.Reference(),
+		Clusters: []types.KmipClusterInfo{
+			{
+				ClusterId: types.KeyProviderId{
+					Id: providerID,
+				},
+			},
+		},
+	}
+	res, err := methods.RetrieveKmipServersStatus_Task(ctx, m.Client(), &req)
+	if err != nil {
+		return nil, err
+	}
+
+	task := NewTask(m.Client(), res.Returnval)
+	taskInfo, err := task.WaitForResultEx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if taskInfo.Result == nil {
+		return nil, nil
+	}
+	result, ok := taskInfo.Result.(types.ArrayOfCryptoManagerKmipClusterStatus)
+	if !ok {
+		return nil, nil
+	}
+	if len(result.CryptoManagerKmipClusterStatus) == 0 {
+		return nil, nil
+	}
+
+	return &result.CryptoManagerKmipClusterStatus[0], nil
```

Comment:
There can only be one here since we already know it's not zero (check up above) we are querying a specific provider ID, and the provider ID is unique per cluster.